specs/phase1/beacon-chain.md

```diff
@@ -166,246 +127,78 @@ class AttestationData(Container):
-    chunk_challenge_responses: List[CustodyChunkResponse, MAX_CUSTODY_CHUNK_CHALLENGE_RESPONSES]
-    custody_key_reveals: List[CustodyKeyReveal, MAX_CUSTODY_KEY_REVEALS]
-    early_derived_secret_reveals: List[EarlyDerivedSecretReveal, MAX_EARLY_DERIVED_SECRET_REVEALS]
-    custody_slashings: List[SignedCustodySlashing, MAX_CUSTODY_SLASHINGS]
-    # Shards
-    shard_transitions: Vector[ShardTransition, MAX_SHARDS]
-    # Light clients
-    light_client_bits: Bitvector[LIGHT_CLIENT_COMMITTEE_SIZE]
-    light_client_signature: BLSSignature
-```
-
-### Extended `BeaconBlock`
-
-Note that the `body` has a new `BeaconBlockBody` definition.
-
-```python
-class BeaconBlock(Container):
-    slot: Slot
-    proposer_index: ValidatorIndex
-    parent_root: Root
-    state_root: Root
-    body: BeaconBlockBody
-```
-
-#### Extended `SignedBeaconBlock`
-
-Note that the `message` has a new `BeaconBlock` definition.
-
-```python
-class SignedBeaconBlock(Container):
-    message: BeaconBlock
-    signature: BLSSignature
-```
-
-### Extended `BeaconState`
-
-Note that aside from the new additions, `Validator` and `PendingAttestation` have new definitions.
-
-```python
-class BeaconState(Container):
-    # Versioning
-    genesis_time: uint64
-    genesis_validators_root: Root
-    slot: Slot
-    fork: Fork
-    # History
-    latest_block_header: BeaconBlockHeader
-    block_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
-    state_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
-    historical_roots: List[Root, HISTORICAL_ROOTS_LIMIT]
-    # Eth1
-    eth1_data: Eth1Data
-    eth1_data_votes: List[Eth1Data, EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH]
-    eth1_deposit_index: uint64
-    # Registry
-    validators: List[Validator, VALIDATOR_REGISTRY_LIMIT]
-    balances: List[Gwei, VALIDATOR_REGISTRY_LIMIT]
-    # Randomness
-    randao_mixes: Vector[Root, EPOCHS_PER_HISTORICAL_VECTOR]
-    # Slashings
-    slashings: Vector[Gwei, EPOCHS_PER_SLASHINGS_VECTOR]  # Per-epoch sums of slashed effective balances
-    # Attestations
+class BeaconState(phase0.BeaconState):
+    # Updated fields
     previous_epoch_attestations: List[PendingAttestation, MAX_ATTESTATIONS * SLOTS_PER_EPOCH]
     current_epoch_attestations: List[PendingAttestation, MAX_ATTESTATIONS * SLOTS_PER_EPOCH]
-    # Finality
-    justification_bits: Bitvector[JUSTIFICATION_BITS_LENGTH]  # Bit set for every recent justified epoch
-    previous_justified_checkpoint: Checkpoint  # Previous epoch snapshot
-    current_justified_checkpoint: Checkpoint
-    finalized_checkpoint: Checkpoint
-    # Phase 1
+    # New fields
+    current_epoch_pending_shard_headers: List[PendingHeader, MAX_PENDING_HEADERS * SLOTS_PER_EPOCH]
+    previous_epoch_pending_shard_headers: List[PendingHeader, MAX_PENDING_HEADERS * SLOTS_PER_EPOCH]
+    most_recent_confirmed_commitments: Vector[Vector[DataCommitment, SLOTS_PER_EPOCH], MAX_SHARDS]
```

Comment:
What exactly does this track? It does not include all confirmed commitments (e.g. the ones confirmed through the fast confirmation mechanism are missing). Should it be the confirmed commitments from previous previous epoch? Should be renamed accordingly

Reply:
It's the confirmed commitments fro the previous previous epoch. Do you have a suggestion for the best name for this? I struggled with naming here...

Reply:
grandfather_epoch_confirmed_commitments?

To be honest though, is this the right thing? Basically it means that a commitment confirmed now (through the fast commitment algorithm) will only be considered for the data availability fork choice two epochs later. I guess that's just in time to prevent finality, but why not consider it earlier (and prevent it from being finalized)?

After all, if we allow it being finalized, that means the chain is in a potentially unrecoverable state (without social intervention) if the data truly isn't available. So shouldn't we make it part of FCR immediately, so that validators will stop voting for that branch?

Reply:
`grandparent` would be probably more in line with `parent`